# wPGSA

Python script used for wPGSA method to estimate relative activities of transcriptional regulators from transcriptome data.

## Introduction

Predicting responsible transcription regulators for given transcriptome data is one of the most promising computational approaches in understanding cellular processes and characteristics. To incorporate information about heterogeneous frequencies of transcription factor (TF)-binding events, we have developed a flexible framework for gene set analysis employing the weighted t-test procedure, namely weighted parametric gene set analysis (wPGSA). Using transcriptome data as an input, wPGSA predicts the activities of transcription regulators responsible for observed gene expression.

## Requirements

- Python
- Python modules: numpy, scipy, rpy2
- R

## Usage

To run wPGSA, execute `wPGSA.py` with logFC expression data and network file;

```
$ python wPGSA.py --logfc-file [logFC_file] --network-file [network_file]
```

To run hierarchical clustering and generate json objects, `hclust.py` with z-score data generated by wPGSA;

```
$ python hclust.py [z_score_data] > data.js
```

## Docker container

Docker container to run wPGSA and hclust of output is [available on Docker hub](https://hub.docker.com/r/inutano/wpgsa/);

```
# prepare logFC file and network file on current directory
$ ls
sample_logFC.txt mm9_sample.network
$ docker run -it -v `pwd`:/data inutano/wpgsa wpgsa --logfc-file /data/sample_logFC.txt --network mm9_sample.network
$ ls
sample_logFC.txt sample_logFC_TF_wPGSA_p_value.txt sample_logFC_TF_wPGSA_q_value.txt sample_logFC_TF_wPGSA_z_score.txt mm9_sample.network
$ docker run -it -v `pwd`:/data inutano/wpgsa hclust /data/sample_logFC_TF_wPGSA_z_score.txt > data.js
```
